script/testing/util/test_server.py

```diff
@@ -1,162 +1,207 @@
 
-        # run the post test tasks
-        test_case.run_post_test()
-        self.db_instance.delete_wal()
+            test_case.run_post_test()
+            self.db_instance.delete_wal()
 
         return ret_val
 
     def run(self, test_suite):
-        """ Orchestrate the overall test execution """
+        """
+        Orchestrate the overall execution of the specified test suite.
+
+        Parameters
+        ----------
+        test_suite : [TestCase]
+            A list of test cases to be run.
+
+        Returns
+        -------
+        suite_result : constants.ErrorCode
+            SUCCESS if all the tests succeeded. ERROR otherwise.
+        """
+        # TODO(WAN): I dislike this magic...
         if not isinstance(test_suite, List):
             test_suite = [test_suite]
 
+        result = constants.ErrorCode.ERROR
         try:
             self.run_pre_suite()
-
-            test_suite_ret_vals = self.run_test_suite(test_suite)
-            test_suite_result = self.determine_test_suite_result(test_suite_ret_vals)
+            exit_codes = self.run_test_suite(test_suite)
+            has_error = any([x is None or x != constants.ErrorCode.SUCCESS
+                             for x in exit_codes.values()])
+            result = constants.ErrorCode.ERROR if has_error else constants.ErrorCode.SUCCESS
         except:
             traceback.print_exc(file=sys.stdout)
-            test_suite_result = constants.ErrorCode.ERROR
         finally:
-            # after the test suite finish, stop the database instance
-            self.db_instance.stop_db()
-        return self.handle_test_suite_result(test_suite_result)
+            self.run_post_suite()
+
+        if result is None or result != constants.ErrorCode.SUCCESS:
+            LOG.error("The test suite failed")
+        return result
 
     def run_test_suite(self, test_suite):
-        """ Execute all the tests in the test suite """
-        test_suite_ret_vals = {}
+        """
+        Run the provided test suite.
+
+        Parameters
+        ----------
+        test_suite : [TestCase]
+            A list of test cases to be run.
+
+        Returns
+        -------
+        exit_codes : dict
+            A dictionary mapping test cases to their exit codes.
+        """
+        exit_codes = {}
         for test_case in test_suite:
+            dbms_started = False
             try:
-                # catch the exception from run_db(), stop_db(), and restart_db()
-                # in case the db is unable to start/stop/restart
-                if test_case.db_restart:
-                    self.db_instance.restart_db(self.is_dry_run)
-                elif not self.db_instance.db_process:
-                    self.db_instance.run_db(self.is_dry_run)
+                if test_case.db_restart and self.db_instance.db_process:
+                    self.db_instance.stop_db(self.is_dry_run)
+                dbms_started = self.db_instance.run_db(self.is_dry_run)
```

Comment:
See comment at end of file for an explanation as to why this needs to be wrapped in an if.